### PR TITLE
Fix driver EFI subsystem for VS and RISCV64 and add driver testing

### PR DESCRIPTION
**Pull request by pbatard**
This series of patches aims at adding driver testing to our test suite.

To accomplish that, we need to fix the subsystem that was improperly set for the Visual Studio driver, as well as fix subsystem generation for RISCV64, which was broken per #27.

To accomplish the latter, we simply set the subsystem using `dd` after the `objcopy` step (which is made a little more complex due to the fact that GNU Make's echo doesn't support `-ne` but still relatively straightforward).

With these patches, we can then add a new test that validates that `drv0` loads and can be invoked.

### CODE REVIEW

File: Make.rules [hunk 1]
```diff
@@ -36,19 +36,27 @@
...
+	$(HIDE)/bin/echo -ne "$(CHAR_SUBSYSTEM)" | dd of=$@ bs=1 seek=156 count=1 conv=notrunc status=none
+endif
 
 %.efi.debug: %.so
 	@$(ECHO) "  OBJCOPY  $(notdir $@)"
 	$(HIDE)$(OBJCOPY) -j .debug_info -j .debug_abbrev -j .debug_aranges \
 		-j .debug_line -j .debug_str -j .debug_ranges \
 		-j .note.gnu.build-id \
 		$(FORMAT) $*.so $@
+ifeq ($(ARCH),riscv64)
```
Comment (gmbr3) [inline Make.rules:57]: debug files don't contain PE header

**Review by gmbr3 — CHANGES_REQUESTED** [review]

File: Make.rules [hunk 2]
```diff
@@ -36,19 +36,27 @@
...
 
+CHAR_SUBSYSTEM=$(subst 0x,\x,$(SUBSYSTEM))
+
 ifeq ($(IS_MINGW32),)
 %.efi: %.so
 	@$(ECHO) "  OBJCOPY  $(notdir $@)"
 	$(HIDE)$(OBJCOPY) -j .text -j .sdata -j .data -j .dynamic -j .rodata -j .rel \
 		    -j .rela -j .rel.* -j .rela.* -j .rel* -j .rela* \
 		    -j .areloc -j .reloc $(FORMAT) $*.so $@
+ifeq ($(ARCH),riscv64)
```
Comment (gmbr3) [inline Make.rules:47]: This could do with being a tool e.g. installed script
It'll break things that don't use generator mode (e.g.fwupd-efi)

**Review by gmbr3 — CHANGES_REQUESTED** [review]

File: Make.rules (hunk 1 above)
Comment (pbatard) [reply Make.rules:57]: OK, Ill fix that. I actually have no idea what these `.efi.debug` files are for...

Comment (gmbr3) [reply Make.rules:57]: > OK, Ill fix that. I actually have no idea what these `.efi.debug` files are for...

GDB - not that I've ever tried... The process to do it looks complicated
The ctors broke I disassembled instead cos it seemed easier than figuring it out

File: Make.rules (hunk 2 above)
Comment (pbatard) [reply Make.rules:47]: I'm not sure I understand what you suggest by "installed script". And I'm not sure how much we want to go the extra mile to cater for downstream projects, when our job is to produce working `.efi` executables by whatever means necessary, which is what downstream should expect, and which is exactly what we are doing here.

Now I am open to suggestions that can help dowstream, but you will have to be very explicit about what you want here, as I am not planning to be looking at fwupd-efi or other downstream projects myself, to figure out how fixing our mandate is going to conflict with stuff they might be relying on... So if you could elaborate on how this could be altered, I would appreciate it.

Comment (gmbr3) [reply Make.rules:47]: > I'm not sure I understand what you 

Having a rethink :)
TLDR: Don't worry


fwupd-efi doesn't need it, they run generator mode as I wrote the code for that. Really, I wished everyone would copy my idea from there as then crt0 can be altered easily without us being worried about breakages. 

Shim is moving to objcopy mode which won't need it either
*this may also make this complicated later but hey ho, I have an idea

fwupd-efi is the only current *live* downstream for RISCV64 (which as I said is unaffected per my code)

Future downstreams will need to make changes anyways for new platforms so they can adapt to this

**Comment by pbatard** [comment]
Okay, I removed the extra step from `%.efi.debug`.

As to getting #27 to be closed automatically, I believe the 2nd commit should do that (as I think it's better to have the issue closed by the commit that fixed it than a PR patchset with multiple commits, including unrelated ones).

**Review by gmbr3 — APPROVED** [review]
LGTM 👍

**Review by ncroxon — APPROVED** [review]
